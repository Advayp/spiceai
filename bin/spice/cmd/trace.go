/*
Copyright 2024-2025 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package cmd

import (
	"fmt"
	"log/slog"

	"github.com/spf13/cobra"
	"github.com/spiceai/spiceai/bin/spice/pkg/context"
	"github.com/spiceai/spiceai/bin/spice/pkg/taskhistory"
)

var (
	// The id of the trace to provide
	id string

	// The trace_id of the trace to provide
	trace_id string
)

var supported_trace_tasks = []string{"ai_chat", "ai_completion", "sql_query", "nsql",
	"tool_use::document_similarity", "tool_use::list_datasets", "tool_use::sql",
	"tool_use::table_schema", "tool_use::sample_data", "tool_use::sql_query", "tool_use::memory"}

func isValidTraceTask(task string) bool {
	for _, supported_task := range supported_trace_tasks {
		if task == supported_task {
			return true
		}
	}
	return false
}

var traceCmd = &cobra.Command{
	Use:   "trace",
	Short: "Return a user friendly trace into an operation that occurred in Spice",
	Example: `
# returns the last trace
$ spice trace ai_chat

# returns the trace for the given id
$ spice trace ai_chat --id chatcmpl-At6ZmDE8iAYRPeuQLA0FLlWxGKNnM
`,
	Args: cobra.MinimumNArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		rtcontext := context.NewContext()
		apiKey, _ := cmd.Flags().GetString("api-key")
		if apiKey != "" {
			rtcontext.SetApiKey(apiKey)
		}

		var filter string
		var err error
		switch isValidTraceTask(args[0]) {
		case true:
			filter, err = getTraceFilter(args[0], id, trace_id)
		default:
			err = fmt.Errorf("invalid trace type %s", args[0])
		}
		if err != nil {
			cmd.Println(err)
			return
		}

		traces, err := taskhistory.SqlRequestToTraces(rtcontext, fmt.Sprintf("SELECT * FROM runtime.task_history WHERE %s ORDER BY start_time asc", filter))
		if err != nil {
			slog.Error("SQL query to 'task_history' failed", "error", err)
			cmd.PrintErrln("Error: failed to retrieve events from runtime.")
			return
		}
		if len(traces) == 0 {
			cmd.PrintErrln("Error: No events found")
			return
		}
		taskhistory.PrintTreeFromTraces(cmd.OutOrStdout(), traces, Display)
	},
}

func Display(t *taskhistory.TaskHistory) string {
	return fmt.Sprintf("(%8.2fms) %s ", t.ExecutionDurationMs, t.Task)
}

func init() {
	RootCmd.AddCommand(traceCmd)
	traceCmd.Flags().StringVar(&id, "id", "", "Return the trace with the given id")
	traceCmd.Flags().StringVar(&trace_id, "trace-id", "", "Return the trace with the given trace id")
}

func getTraceFilter(task string, id string, trace_id string) (string, error) {
	if id != "" {
		return fmt.Sprintf("trace_id=(SELECT trace_id from runtime.task_history where labels.id='%s')", id), nil
	}
	if trace_id != "" {
		return fmt.Sprintf("trace_id='%s'", trace_id), nil
	}
	// use last by default
	return fmt.Sprintf("trace_id=(SELECT trace_id from runtime.task_history where task='%s' order by start_time desc limit 1)", task), nil
}
