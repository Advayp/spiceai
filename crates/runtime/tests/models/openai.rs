/*
Copyright 2024 The Spice.ai OSS Authors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use app::AppBuilder;
use async_openai::types::{
    ChatCompletionRequestSystemMessageArgs, ChatCompletionRequestUserMessageArgs,
    CreateChatCompletionRequestArgs, EmbeddingInput,
};
use std::sync::Arc;

use llms::chat::Chat;
use opentelemetry_sdk::trace::TracerProvider;
use runtime::{auth::EndpointAuth, model::try_to_chat_model, Runtime};
use spicepod::component::{
    embeddings::{ColumnEmbeddingConfig, Embeddings},
    model::Model,
};

use crate::{
    init_tracing, init_tracing_with_task_history,
    models::{
        create_api_bindings_config, get_executed_tasks, get_params_with_secrets,
        get_taxi_trips_dataset, get_tpcds_dataset, json_is_single_row_with_value,
        normalize_chat_completion_response, normalize_embeddings_response,
        normalize_search_response, pretty_json_str, send_chat_completions_request,
        send_nsql_request, send_search_request,
    },
    utils::{runtime_ready_check, verify_env_secret_exists},
};

use super::send_embeddings_request;

#[tokio::test]
async fn openai_test_nsql() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_model(get_openai_model("gpt-4o-mini", "nql"))
        .with_model(get_openai_model("gpt-4o-mini", "nql-2"))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);

    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let (_tracing, trace_provider) = init_tracing_with_task_history(None, &rt);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    // example responses for the test query: '[{"count(*)":10}]', '[{"record_count":10}]'

    tracing::info!("/v1/nsql: Verify default nsql request");
    let task_start_time = std::time::SystemTime::now();

    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        None,
        Some(false),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let table_schema_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::table_schema")
        .expect("Expected 'tool_use::table_schema' task to be executed");

    insta::assert_snapshot!(
        "nsql_table_schema_task",
        pretty_json_str(&table_schema_task.1)?
    );

    tracing::info!("/v1/nsql: Ensure model selection works");
    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        Some("nql-2"),
        Some(false),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    tracing::info!("/v1/nsql: Verify nsql request with 'sample_data_enabled:true'");

    let task_start_time = std::time::SystemTime::now();

    let response = send_nsql_request(
        http_base_url.as_str(),
        "how many records in taxi_trips dataset?",
        Some("nql"),
        Some(true),
        None,
    )
    .await?;

    assert!(
        json_is_single_row_with_value(&response, 10),
        "Expected a single record containing the value 10"
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let sample_data_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::sample_data")
        .expect("Expected 'tool_use::sample_data' task to be executed");

    insta::assert_snapshot!("nsql_sample_data_task", sample_data_task.1);

    let sql_query_task = tasks
        .iter()
        .find(|t| t.0 == "sql_query")
        .expect("Expected 'sql_query' task to be executed");

    insta::assert_snapshot!("nsql_sample_data_task_sql_query", sql_query_task.1);

    Ok(())
}

#[tokio::test]
async fn openai_test_search() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut ds_tpcds_item = get_tpcds_dataset("item");
    ds_tpcds_item.embeddings = vec![ColumnEmbeddingConfig {
        column: "i_item_desc".to_string(),
        model: "openai_embeddings".to_string(),
        primary_keys: Some(vec!["i_item_sk".to_string()]),
        chunking: None,
    }];

    let app = AppBuilder::new("search_app")
        // taxi_trips dataset is used to test search when there is a dataset w/o embeddings
        .with_dataset(get_taxi_trips_dataset())
        .with_dataset(ds_tpcds_item)
        .with_embedding(get_openai_embeddings(
            Option::<String>::None,
            "openai_embeddings",
        ))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    tracing::info!("/v1/search: Ensure simple search request succeeds");
    let response = send_search_request(
        http_base_url.as_str(),
        "new patient",
        Some(2),
        Some(vec!["item".to_string()]),
        None,
        Some(vec!["i_color".to_string(), "i_item_id".to_string()]),
    )
    .await?;

    insta::assert_snapshot!(format!("search_1"), normalize_search_response(response));

    tracing::info!("/v1/search: Ensure search request across all datasets succeeds");
    let response = send_search_request(
        http_base_url.as_str(),
        "new patient",
        Some(2),
        None,
        None,
        None,
    )
    .await?;

    insta::assert_snapshot!(format!("search_2"), normalize_search_response(response));

    Ok(())
}

#[tokio::test]
async fn openai_test_embeddings() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let app = AppBuilder::new("search_app")
        .with_embedding(get_openai_embeddings(
            Some("text-embedding-3-small"),
            "openai_embeddings",
        ))
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    let embeddins_test = vec![
        (
            EmbeddingInput::String("The food was delicious and the waiter...".to_string()),
            Some("float"),
            None,
            None,
        ),
        (
            EmbeddingInput::StringArray(vec![
                "The food was delicious".to_string(),
                "and the waiter...".to_string(),
            ]),
            None,
            Some("test_user_id"),
            Some(256),
        ),
        (
            EmbeddingInput::StringArray(vec![
                "The food was delicious".to_string(),
                "and the waiter...".to_string(),
            ]),
            Some("base64"),
            None,
            Some(128),
        ),
    ];

    let mut test_id = 0;

    for (input, encoding_format, user, dimensions) in embeddins_test {
        test_id += 1;
        let response = send_embeddings_request(
            http_base_url.as_str(),
            "openai_embeddings",
            input,
            encoding_format,
            user,
            dimensions,
        )
        .await?;

        insta::assert_snapshot!(
            format!("embeddings_{}", test_id),
            // OpenAI's embeddings response is not deterministic (values vary for the same input, model version, and parameters) so
            // we normalize the response before snapshotting
            normalize_embeddings_response(response)
        );
    }

    Ok(())
}

#[tokio::test]
async fn openai_test_chat_completion() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut model_with_tools = get_openai_model("gpt-4o-mini", "openai_model");
    model_with_tools
        .params
        .insert("tools".to_string(), "auto".into());

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_model(model_with_tools)
        .build();

    let api_config = create_api_bindings_config();
    let http_base_url = format!("http://{}", api_config.http_bind_address);
    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let rt_ref_copy = Arc::clone(&rt);
    tokio::spawn(async move {
        Box::pin(rt_ref_copy.start_servers(api_config, None, EndpointAuth::no_auth())).await
    });

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(10)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    let response = send_chat_completions_request(
        http_base_url.as_str(),
        vec![
            ("system".to_string(), "You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string()),
            ("user".to_string(), "Provide the total number of records in the taxi trips dataset. If known, return a single numeric value.".to_string()),
        ],
        "openai_model",
        false,
    ).await?;

    insta::assert_snapshot!(
        "chat_completion",
        normalize_chat_completion_response(response, false)
    );

    Ok(())
}

#[tokio::test]
async fn openai_test_chat_messages() -> Result<(), anyhow::Error> {
    let _tracing = init_tracing(None);

    verify_env_secret_exists("SPICE_OPENAI_API_KEY")
        .await
        .map_err(anyhow::Error::msg)?;

    let mut ds_tpcds_item = get_tpcds_dataset("item");
    ds_tpcds_item.embeddings = vec![ColumnEmbeddingConfig {
        column: "i_item_desc".to_string(),
        model: "openai_embeddings".to_string(),
        primary_keys: Some(vec!["i_item_sk".to_string()]),
        chunking: None,
    }];

    let app = AppBuilder::new("text-to-sql")
        .with_dataset(get_taxi_trips_dataset())
        .with_dataset(ds_tpcds_item)
        .with_embedding(get_openai_embeddings(
            Some("text-embedding-3-small"),
            "openai_embeddings",
        ))
        .build();

    let rt = Arc::new(Runtime::builder().with_app(app).build().await);

    let (_tracing, trace_provider) = init_tracing_with_task_history(None, &rt);

    tokio::select! {
        () = tokio::time::sleep(std::time::Duration::from_secs(30)) => {
            return Err(anyhow::anyhow!("Timed out waiting for components to load"));
        }
        () = rt.load_components() => {}
    }

    runtime_ready_check(&rt).await;

    verify_sql_query_chat_completion(Arc::clone(&rt), &trace_provider).await?;
    verify_similarity_search_chat_completion(Arc::clone(&rt), &trace_provider).await?;

    Ok(())
}

/// Verifies that the model correctly uses the SQL tool to process user query and return the result
async fn verify_sql_query_chat_completion(
    rt: Arc<Runtime>,
    trace_provider: &TracerProvider,
) -> Result<(), anyhow::Error> {
    let model =
        get_openai_chat_model(Arc::clone(&rt), "gpt-4o-mini", "openai_model", "auto").await?;
    let req = CreateChatCompletionRequestArgs::default()
            .messages(vec![ChatCompletionRequestSystemMessageArgs::default()
                .content("You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string())
                .build()?
                .into(),ChatCompletionRequestUserMessageArgs::default()
                .content("Provide the total number of records in the taxi trips dataset. If known, return a single numeric value.".to_string())
                .build()?
                .into()])
            .build()?;

    let task_start_time = std::time::SystemTime::now();
    let response = model.chat_request(req).await?;

    insta::assert_snapshot!(
        "chat_1_response_choices",
        format!("{:#?}", response.choices)
    );

    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    assert!(
        tasks.iter().any(|t| { t.0 == "tool_use::list_datasets" }),
        "Expected 'tool_use::list_datasets' task to be executed"
    );
    assert!(
        tasks.iter().any(|t| { t.0 == "tool_use::sql" }),
        "Expected 'tool_use::sql' task to be executed"
    );
    assert!(
        tasks.iter().any(|t| { t.0 == "sql_query" }),
        "Expected 'sql_query' task to be executed"
    );

    let ai_completion_task =
        tasks
            .iter()
            .find(|t| t.0 == "ai_completion")
            .ok_or(anyhow::anyhow!(
                "Expected 'ai_completion' task to be executed"
            ))?;

    // ai_completion input message is deterministic - based on available tools, app configuration, and the input message
    insta::assert_snapshot!(
        "chat_1_ai_completion_input",
        pretty_json_str(&ai_completion_task.1)?
    );

    Ok(())
}

/// Verifies that the model correctly uses similirity search tool to process user query and return the result
async fn verify_similarity_search_chat_completion(
    rt: Arc<Runtime>,
    trace_provider: &TracerProvider,
) -> Result<(), anyhow::Error> {
    let model =
        get_openai_chat_model(Arc::clone(&rt), "gpt-4o-mini", "openai_model", "auto").await?;
    let req = CreateChatCompletionRequestArgs::default()
                .messages(vec![ChatCompletionRequestSystemMessageArgs::default()
                    .content("You are an assistant that responds to queries by providing only the requested data values without extra explanation.".to_string())
                    .build()?
                    .into(),ChatCompletionRequestUserMessageArgs::default()
                    .content("Find information about vehicles and journalists".to_string())
                    .build()?
                    .into()])
                .build()?;

    let task_start_time = std::time::SystemTime::now();
    let mut response = model.chat_request(req).await?;

    // Verify model correctly found records in the dataset
    assert!(response.choices[0]
        .message
        .content
        .clone()
        .unwrap_or_default()
        .contains("there just big vehicles. Journalists"));

    // response contains distance metrics that can vary so we replace content with a placeholder for snapshotting
    response.choices.iter_mut().for_each(|c| {
        c.message.content = Some("__placeholder__".to_string());
    });

    insta::assert_snapshot!(
        "chat_2_response_choices",
        format!("{:#?}", response.choices)
    );

    // ensure all spans are exported into task_history
    let _ = trace_provider.force_flush();

    let tasks = get_executed_tasks(&rt, task_start_time.into()).await?;

    let document_similarity_task = tasks
        .iter()
        .find(|t| t.0 == "tool_use::document_similarity")
        .ok_or(anyhow::anyhow!(
            "Expected 'document_similarity' task to be executed"
        ))?;

    // ai_completion input message is deterministic - based on available tools, app configuration, and the input message
    insta::assert_snapshot!(
        "chat_2_document_similarity_input",
        pretty_json_str(&document_similarity_task.1)?
    );

    Ok(())
}

fn get_openai_model(model: impl Into<String>, name: impl Into<String>) -> Model {
    let mut model = Model::new(format!("openai:{}", model.into()), name);
    model.params.insert(
        "openai_api_key".to_string(),
        "${ secrets:SPICE_OPENAI_API_KEY }".into(),
    );
    model
}

async fn get_openai_chat_model(
    rt: Arc<Runtime>,
    model: impl Into<String>,
    name: impl Into<String>,
    tools: impl Into<String>,
) -> Result<Box<dyn Chat>, anyhow::Error> {
    let mut model_with_tools = get_openai_model(model, name);
    model_with_tools
        .params
        .insert("tools".to_string(), tools.into().into());

    let model_secrets = get_params_with_secrets(&model_with_tools.params, &rt).await;
    try_to_chat_model(&model_with_tools, &model_secrets, rt)
        .await
        .map_err(anyhow::Error::from)
}

fn get_openai_embeddings(model: Option<impl Into<String>>, name: impl Into<String>) -> Embeddings {
    let mut embedding = match model {
        Some(model) => Embeddings::new(format!("openai:{}", model.into()), name),
        None => Embeddings::new("openai", name),
    };
    embedding.params.insert(
        "openai_api_key".to_string(),
        "${ secrets:SPICE_OPENAI_API_KEY }".into(),
    );
    embedding
}
