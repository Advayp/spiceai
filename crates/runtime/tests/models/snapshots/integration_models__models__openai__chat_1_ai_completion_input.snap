---
source: crates/runtime/tests/models/openai.rs
expression: "sql_to_display(&rt,\nformat!(r#\"SELECT input\n                FROM runtime.task_history\n                WHERE start_time >= '{}'\n                AND task='ai_completion'\n                ORDER BY start_time\n                LIMIT 1;\n            \"#,\nInto::<DateTime<Utc>>::into(task_start_time).to_rfc3339()).as_str()).await.expect(\"Failed to execute HTTP SQL query\")"
---
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| input                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| {"messages":[{"role":"assistant","tool_calls":[{"id":"initial_list_datasets","type":"function","function":{"name":"list_datasets","arguments":""}}]},{"role":"tool","content":"[{\"table\":\"spice.public.taxi_trips\",\"can_search_documents\":false,\"description\":null,\"metadata\":{}},{\"table\":\"spice.public.item\",\"can_search_documents\":true,\"description\":null,\"metadata\":{}}]","tool_call_id":"initial_list_datasets"},{"role":"system","content":"You are an assistant that responds to queries by providing only the requested data values without extra explanation."},{"role":"user","content":"Provide the total number of records in the taxi trips dataset. If known, return a single numeric value."}],"model":"","tools":[{"type":"function","function":{"name":"document_similarity","description":"Search and retrieve documents from available datasets","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"SearchRequestJson","type":"object","required":["text"],"properties":{"additional_columns":{"description":"Additional columns to return from the dataset.","default":[],"type":"array","items":{"type":"string"}},"datasets":{"description":"The datasets to search for similarity. If None, search across all datasets. For available datasets, use the `list_datasets` tool and ensure `can_search_documents==true`.","default":null,"type":["array","null"],"items":{"type":"string"}},"limit":{"description":"Number of documents to return for each dataset","default":null,"type":["integer","null"],"format":"uint","minimum":0.0},"text":{"description":"The text to search documents for similarity","type":"string"},"where":{"description":"An SQL filter predicate to apply. Format: 'WHERE `where_cond`'.","default":null,"type":["string","null"]}}}}},{"type":"function","function":{"name":"get_readiness","description":"Retrieves the readiness status of all runtime components including registered datasets, models, and embeddings."}},{"type":"function","function":{"name":"list_datasets","description":"List all SQL tables available."}},{"type":"function","function":{"name":"random_sample","description":"Sample random rows from a table.","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"RandomSampleParams","type":"object","required":["dataset","limit"],"properties":{"dataset":{"description":"The SQL dataset to sample data from.","type":"string"},"limit":{"description":"The number of rows, each with distinct values per column, to sample.","type":"integer","format":"uint","minimum":0.0}}}}},{"type":"function","function":{"name":"sample_distinct_columns","description":"Generate synthetic data by sampling distinct column values from a table.","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"DistinctColumnsParams","type":"object","required":["dataset","limit"],"properties":{"cols":{"description":"The columns to sample distinct values from. If None, all columns are sampled.","type":["array","null"],"items":{"type":"string"}},"dataset":{"description":"The SQL dataset to sample data from.","type":"string"},"limit":{"description":"The number of rows, each with distinct values per column, to sample.","type":"integer","format":"uint","minimum":0.0}}}}},{"type":"function","function":{"name":"sql","description":"Run an SQL query on the data source. Columns with capitals must be quoted. When needed quote each part of catalog.schema.table: \"catalog\".\"schema\".\"table\". Avoid 'SELECT *', and columns with `_offset` or `_embedding` suffix.","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"SqlToolParams","type":"object","required":["query"],"properties":{"query":{"description":"The SQL query to run. Double quote all select columns and never select columns ending in '_embedding'. The `table_catalog` is 'spice'. Always use it in the query","type":"string"}}}}},{"type":"function","function":{"name":"table_schema","description":"Retrieve the schema of all available SQL tables","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"TableSchemaToolParams","type":"object","required":["tables"],"properties":{"output":{"description":"If `full` return metadata and semantic details about the columns.","default":"full","allOf":[{"$ref":"#/definitions/OutputType"}]},"tables":{"description":"Which subset of tables to return results for. Default to all tables.","type":"array","items":{"type":"string"}}},"definitions":{"OutputType":{"type":"string","enum":["full","minimal"]}}}}},{"type":"function","function":{"name":"top_n_sample","description":"Sample the top N rows from a table based on a specified ordering","parameters":{"$schema":"http://json-schema.org/draft-07/schema#","title":"TopSamplesParams","type":"object","required":["dataset","limit","order_by"],"properties":{"dataset":{"description":"The SQL dataset to sample data from.","type":"string"},"limit":{"description":"The number of rows, each with distinct values per column, to sample.","type":"integer","format":"uint","minimum":0.0},"order_by":{"description":"How to order the samples before retrieving the top N.","type":"string"}}}}}]} |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
